File: TobiiExampleScript/Tobii_helpers/t2t/examples/python/cmdTest.py
#!/usr/bin/env python

"""
# Luca Filippin - July 2010 - luca.filippin@gmail.com                                                
# Copyright 2010 Language, Cognition and  Development Laboratory, Sissa, Trieste. All rights reserved.
# 
# PLEASE HAVE A LOOK TO THE SWIG DOCUMENTATION FOR A DEEPER UNDERSTANDING OF THIS CODE
# NOTE: Swig wrapper to t2t makes use of shadow classes and swig helpers (see t2tsw.i)
"""

from argparse import ArgumentParser
from t2tsw import *
from t2tHelpers import *
import sys
import os
import time
import re

def banner():
    print ""
    print "+-------------------------------------------------------------------------------------+"
    print "|  This is a test application which tests some Tobii ET commands.                     |"
    print "|-------------------------------------------------------------------------------------|"
    print "|  Luca Filippin - July 2010 - luca.filippin@gmail.com                                |"                                                
    print "|  Copyright 2010 Language, Cognition and  Development Laboratory, Sissa, Trieste     |"
    print "+-------------------------------------------------------------------------------------+"
    print ""

def parseOptions():
    parser = ArgumentParser(description = 'Test a Tobii Eye Tracker')
    parser.add_argument('tobii_ip', help='tobii ip address')
    parser.add_argument('--version', action = 'version', version="%(prog)s v1.0")
    parser.add_argument("-l", "--log-file", dest="tobii_log", default="", help="log file name")
    parser.add_argument("-c", "--calibration-file", dest="calibration_file", default="", help="Use a stored calibration file")
    parser.add_argument("-t", "--tracked-data-file", dest="tracked_data_file", default="EyesTrackedData.txt", help="File where tracked data will be stored in the TET machine")
    parser.add_argument("-e", "--events-data-file", dest="events_data_file", default="EventsData.txt", help="File where events will be stored in the TET machine")
    parser.add_argument("-p", "--port", dest="tobii_port", type=int, default=4455, help="TET server listening port")
   
    try:
        o = parser.parse_args()
        if o.tobii_port < 0:
            raise Exception("Bad port value '%(v)s': should be > 0" %{ 'v' : o.tobii_port })
    except Exception, e:
        raise Exception("++ Argument error\n" + e.__str__())
    
    return o

try:

    banner()
    tobii_connected = False
    o = parseOptions() 
    if o.tobii_log <> "": 
        t2tOutputFileName(o.tobii_log, "w")
    
    # Connect to the tobii
    print "Connecting to the TET server @" + o.tobii_ip + ":" + '%(p)d' %{ 'p':o.tobii_port } + "..."
    c = t2tCmd()
    c.cmd = "CONNECT"
    c.prm.connect.ip_address = o.tobii_ip
    c.prm.connect.port = o.tobii_port
    cmdEx(c)
    checkStatus("connect", 1, timeout = 5, debug=False)
    if not checkStatus("connected", 1, timeout = 5, debug=False): 
        bailOut(msg = "Connection failed\nExiting", disconnect = False)
    tobii_connected = True
    
    raw_input("\n***** PRESS <enter> TO START THE TEST *****\n")
    
    # Run the demo 
    print "Running demo..."
    c = t2tCmd()
    c.cmd = "DEMO"
    cmdEx(c)
    
    if o.calibration_file <> "":
        print "Loading calibration data " + o.calibration_file + "..."
        c = t2tCmd()
        c.cmd = "START_CALIBRATION"
        c.prm.start_calibration.load_from_file = 1
        c.prm.start_calibration.fname = o.calibration_file
        c.prm.start_calibration.cmatrix.cols = 0
        c.prm.start_calibration.cmatrix.rows = 0
        c.prm.start_calibration.cmatrix.vals = None
        cmdEx(c)
        
        checkStatus("calibrating", 1, timeout = 5)
        if not checkStatus("calibstarted", 1, timeout = 5): 
            bailOut(msg = "Start calibration failed\nExiting", disconnect = True)
        
        # there's no calibend flag set for loading
        if not checkStatus("calibrating", 0, timeout = 20): 
            bailOut(msg = "Start calibration failed\nExiting", disconnect = True)
        
        c = t2tCmd()
        c.cmd = "CALIBRATION_ANALYSIS"  
        calib_an = cmdEx(c).calibration_analysis
        print "\nCalibration analisys data:\n\n" + t2tCalibrationData._compact_header + "\n" + calib_an.__str__(compact = True)
        
        print "Removing a couple of calibration samples set..."
        c = t2tCmd()
        c.cmd = "REMOVE_CALIBRATION_SAMPLES"
        c.prm.remove_calibration_samples.rmatrix.rows = 2
        c.prm.remove_calibration_samples.rmatrix.cols = 4
        vals = doubleArrayC(4*2)
        c.prm.remove_calibration_samples.rmatrix.vals = vals
        vals[0] = 1       # eye 
        vals[1] = 0.2     # x
        vals[2] = 0.15    # y
        vals[3] = 0.12    # radius
        vals[4] = 3       # eye
        vals[5] = 0.8     # x
        vals[6] = 0.7     # y
        vals[7] = 0.1     # radius
        cmdEx(c)
        
        if not checkStatus("removing_samples", 1, timeout = 5) or not checkStatus("removing_samples", 0, timeout = 5): 
            bailOut(msg = "Calibration samples removal failed\nExiting", disconnect = True)
        
        print "Recalculating & setting calibration..."
        c = t2tCmd()
        c.cmd = "START_CALIBRATION"
        c.prm.start_calibration.load_from_file = 0
        c.prm.start_calibration.fname = None
        c.prm.start_calibration.cmatrix.cols = 0
        c.prm.start_calibration.cmatrix.rows = 0
        c.prm.start_calibration.cmatrix.vals = None
        cmdEx(c)
        
        checkStatus("calibrating", 1, timeout = 5)
        if not checkStatus("calibstarted", 1, timeout = 5): 
            bailOut(msg = "Start calibration failed\nExiting", disconnect = True)


    # Do sync only after start_tracking or it will fail ??? It happens on T60
    print "Synchronizing..."
    c = t2tCmd()
    c.cmd = "SYNCHRONISE"
    cmdEx(c)
    checkStatus("synchronise", 1, timeout = 5)
    if not checkStatus("synchronised", 1, timeout = 5): 
        bailOut(msg = "Synchronizing failed\nExiting", disconnect = True)
    
    print "Start auto-sync..."
    c = t2tCmd()
    c.cmd = "START_AUTO_SYNC"
    cmdEx(c)
    if not checkStatus("autosynced", 1, timeout = 5): 
        bailOut(msg = "Start auto-sync failed\nExiting", disconnect = True)
    
    print "Start tracking..."
    c = t2tCmd()
    c.cmd = "START_TRACKING"
    cmdEx(c)
    checkStatus("running", 1, timeout = 5)
    if not checkStatus("runstarted", 1, timeout = 5): 
        bailOut(msg = "Start tracking failed\nExiting", disconnect = True)
    
    print "Getting some extended samples..."
    for i in xrange(0, 10):
        c = t2tCmd()
        c.cmd = "GET_SAMPLE_EXT"    
        smp = cmdEx(c).sample_ext
        print "> ", smp
        time.sleep(0.2)
    
    print "Getting some timestamps..."
    for i in xrange(0, 10):
        c = t2tCmd()
        c.cmd = None    
        print "> %f" %(cmdEx(c).timestamp.time)
        
    print "Start recording..."
    c = t2tCmd()
    c.cmd = "RECORD"
    cmdEx(c)
    time.sleep(2)
    
    print "Sending events..."
    for i in xrange(0, 5):
        c = t2tCmd()
        c.cmd = "EVENT"
        c.prm.event.name = "Event %d" %(i) 
        c.prm.event.duration = 1
        c.prm.event.nfields = 2
        c.prm.event.fields = new_charpArray(2)
        charpArray_setitem(c.prm.event.fields, 0, "FIELD 1")
        charpArray_setitem(c.prm.event.fields, 1, "FIELD 2")
        vals = doubleArrayC(2)
        c.prm.event.values = vals
        vals[0] = i*10
        vals[1] = i*10 + 1
        start_time = time.time()
        data = cmdEx(c)
        print "> Name = %s local_start = %f secs real_starts = %f secs duration = %s secs" %(c.prm.event.name, start_time, data.event.start_time, c.prm.event.duration)
        print "> Sleeping a bit..."
        delete_charpArray(c.prm.event.fields);
        time.sleep(1.5)
    
    print "Getting gazes data..."
    c = t2tCmd()
    c.cmd = "GET_GAZES_DATA"
    c.prm.get_gazes_data.from_sample_idx = 0
    data = cmdEx(c)
    if data.gazes_data <> None:
        print "> Start time: %f" %data.gazes_data.start_time
        for s in data.gazes_data.samples:
            print "> " +  s.__str__()
    else:
        print "No data"
    
    print "Getting events data..."
    c = t2tCmd()
    c.cmd = "GET_EVENTS_DATA"
    c.prm.get_events_data.from_event_idx = 0
    data = cmdEx(c)
    if data.events_data <> None:
        print "> Start time: %f" %(data.events_data.start_time)
        for e in data.events_data.events:
            print "> " + e.__str__()
    else:
        print "No data"
    
    c =t2tCmd()
    print "Saving data..."
    c.cmd = "SAVE_DATA"
    c.prm.save_data.eye_tracking_fname = o.tracked_data_file   
    c.prm.save_data.events_fname = o.events_data_file 
    c.prm.save_data.mode = "TRUNK"
    cmdEx(c)   
    
    print "Getting status and history..."
    c = t2tCmd()
    c.cmd = "GET_STATUS"
    c.prm.get_status.get_history = 1
    data = cmdEx(c)
    print "> Status: " + data.status_data.__str__()
    for f in data.history_data.facts:
        print "> " + f.__str__()
     
    print "Stop recording..."
    c = t2tCmd()
    c.cmd = "STOP_RECORD"
    cmdEx(c)
    
    print "Stop tracking..."
    c = t2tCmd()
    c.cmd = "STOP_TRACKING"
    cmdEx(c)
    checkStatus("stop", 1, timeout = 1)
    if not checkStatus("running", 0, timeout = 5): 
        bailOut(msg = "Stop tracking failed\nExiting", disconnect = True)
    
    print "Clearing data..."
    c = t2tCmd()
    c.cmd = "CLEAR_DATA"
    c.prm.clear_data.up_sample_index = -2;
    c.prm.clear_data.up_event_index = -2;
    cmdEx(c)
    
    print "Clearing history..."
    c = t2tCmd()
    c.cmd = "CLEAR_HISTORY"
    cmdEx(c)
    
    print "Stop auto-sync..."
    c = t2tCmd()
    c.cmd = "STOP_AUTO_SYNC"
    cmdEx(c)
    if not checkStatus("autosynced", 0, timeout = 5): 
        bailOut(msg = "Stop auto-sync failed\nExiting", disconnect = True)
   
    tobii_connected = False
    print "Disconnecting..."    
    c = t2tCmd()
    c.cmd = "DISCONNECT"
    cmdEx(c, disconnect = False)
    checkStatus("disconnect", 1, timeout = 5)
    if not checkStatus("connected", 0, timeout = 5): 
        bailOut(msg = "Disconnect failed\nExiting", disconnect = False)
    
    print "Cleanup..."
    c = t2tCmd()
    c.cmd = "CLEANUP"
    cmdEx(c)
    
    print "End.\n"
    
    t2tOutputFileName(None, None)
    
except Exception, e:
    msg =  "--------------- UNEXPECTED ERROR ---------------\n" + e.__str__()
    msg += "\n------------------------------------------------\n"
    msg += "\nThe application has been forcely closed!\n"
    bailOut(msg = msg, disconnect = tobii_connected)

File: TobiiExampleScript/Tobii_helpers/t2t/examples/perl/cmdTest.pl
#!/usr/bin/env perl -w
##
# Luca Filippin - July 2010 - luca.filippin@gmail.com                                                
# Copyright 2010 Language, Cognition and  Development Laboratory, Sissa, Trieste. All rights reserved.
# 
# PLEASE HAVE A LOOK TO THE SWIG DOCUMENTATION FOR A DEEPER UNDERSTANDING OF THIS CODE
# NOTE: Swig wrapper to t2t makes use of shadow classes and swig helpers (see t2tsw.i)
##

#use strict;
use Getopt::Long;
use File::Basename;
use Time::HiRes;
use t2tHelpers;

GetOptions(	"help|h"=>\$opt_help,
			"version"=>\$opt_version,
			"log-file|l:s"=>\$opt_log_file,
   			"port|p:i"=>\$opt_tobii_port,
            "calibration-file|c:s"=>\$opt_calibration_file,
            "tracked-data-file|t:s"=>\$opt_data_file,
            "events-data-file|e:s"=>\$opt_evts_file);
            
($command) = fileparse($0);

$version=<<VERSION;
+-------------------------------------------------------------------------------------+
|  cmdTest.pl v1.0                                                                    |
|  This is a test application which tests some Tobii ET commands.                     |
|-------------------------------------------------------------------------------------|
|  Luca Filippin - July 2010 - luca.filippin\@gmail.com                                |                                                
|  Copyright 2010 Language, Cognition and  Development Laboratory, Sissa, Trieste     |
+-------------------------------------------------------------------------------------+
VERSION

$usage=<<USAGE;

Usage: $command [options] <TET ip address>

Test a Tobii Eye Tracker

Options:
  --version             show program's version number and exit
  -h, --help            show this help message and exit
  -l TOBII_LOG, --log-file=TOBII_LOG
                        log file name
  -c CALIBRATION_FILE, --calibration-file=CALIBRATION_FILE
                        Use a stored calibration file
  -t TRACKED_DATA_FILE, --tracked-data-file=TRACKED_DATA_FILE
                        File where tracked data will be stored in the TET
                        machine
  -e EVENTS_DATA_FILE, --events-data-file=EVENTS_DATA_FILE
                        File where events will be stored in the TET machine
  -p TOBII_PORT, --port=TOBII_PORT
                        TET server listening port

USAGE

if ($opt_version) {
	print "$version";
	exit 0;
}

if (@ARGV < 1 or $opt_help) {
  	print "$usage";
  	exit 0;
}

if (!defined($opt_tobii_port)) { $opt_tobii_port=4455; }
if (!defined($opt_start_delay)) { $opt_start_delay=0; }
if (!defined($opt_data_file)) { $opt_data_file="EyesTrackedData.txt"; }
if (!defined($opt_evts_file)) { $opt_evts_file="EventsData.txt"; }

$opt_tobii_ip = $ARGV[0];

sub banner() {
	print "$version\n";
}

$disconnect = 0;

eval {
	banner();
	t2tsw::t2tOutputFileName($opt_log_file, "w") if defined $opt_log_file;
	
	# Connect to the tobii
    print "Connecting to the TET server @".$opt_tobii_ip.":".$opt_tobii_port."...\n";
    my $c = new t2tsw::t2tCmd();
    $c->{cmd} = "CONNECT";
    $c->{prm}->{connect}->{ip_address} = $opt_tobii_ip;
    $c->{prm}->{connect}->{port} = $opt_tobii_port;
    t2tHelpers::cmdEx($c);
	t2tHelpers::checkStatus('connect', 1, 5);
	t2tHelpers::bailOut("Connection failed\nExiting\n", 0) unless t2tHelpers::checkStatus('connected', 1, 5);
	$disconnect = 1;
    
    t2tHelpers::rawInput("\n***** PRESS ANY KEY TO START THE TEST *****\n");
    
    # Running demo
    printf "Running demo...\n";
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "DEMO";
    t2tHelpers::cmdEx($c);
    
    if (defined $opt_calibration_file) {
    	print "Loading calibration data $opt_calibration_file...\n";
    	$c = new t2tsw::t2tCmd();
    	$c->{cmd} = "START_CALIBRATION";
    	$c->{prm}->{start_calibration}->{load_from_file} = 1;
    	$c->{prm}->{start_calibration}->{cmatrix}->{cols} = 0;
    	$c->{prm}->{start_calibration}->{cmatrix}->{rows} = 0;
    	$c->{prm}->{start_calibration}->{cmatrix}->{vals} = undef;
    	$c->{prm}->{start_calibration}->{fname} = $opt_calibration_file;
    	t2tHelpers::cmdEx($c);
    	
    	t2tHelpers::checkStatus('calibrating', 1, 5);
		t2tHelpers::bailOut("Load calibration failed\nExiting\n", 1) unless t2tHelpers::checkStatus('calibstarted', 1, 5);
		t2tHelpers::bailOut("Load calibration failed\nExiting\n", 1) unless t2tHelpers::checkStatus('calibstarted', 0, 5);
    
    	$c = new t2tsw::t2tCmd();
    	$c->{cmd} = "CALIBRATION_ANALYSIS";
    	$calib_an = t2tHelpers::cmdEx($c)->{calibration_analysis};
    
    	if (defined($calib_an)) {
        	print "\nCalibration analysis data:\n\n".$calib_an->compact_header()."\n".$calib_an->str(1)."\n";
        }
        
        print "Removing a couple of calibration samples set...\n";
        $c = new t2tsw::t2tCmd();
        $c->{cmd} = "REMOVE_CALIBRATION_SAMPLES";
        $c->{prm}->{remove_calibration_samples}->{rmatrix}->{rows} = 2;
        $c->{prm}->{remove_calibration_samples}->{rmatrix}->{cols} = 4;
        $c->{prm}->{remove_calibration_samples}->{rmatrix}->{vals} = $vals = new t2tsw::doubleArrayC(2*4);
        t2tsw::doubleArrayC::setitem($vals, 0, 1);       # eye 
        t2tsw::doubleArrayC::setitem($vals, 1, 0.2);     # x
        t2tsw::doubleArrayC::setitem($vals, 2, 0.15);    # y
        t2tsw::doubleArrayC::setitem($vals, 3, 0.12);    # radius
        t2tsw::doubleArrayC::setitem($vals, 4, 3);       # eye 
        t2tsw::doubleArrayC::setitem($vals, 5, 0.8);     # x
        t2tsw::doubleArrayC::setitem($vals, 6, 0.7);     # y
        t2tsw::doubleArrayC::setitem($vals, 7, 0.1);     # radius
		t2tHelpers::cmdEx($c);
		my $remove_ok = (t2tHelpers::checkStatus('removing_samples', 1, 5) && t2tHelpers::checkStatus('removing_samples', 0, 5));
		t2tHelpers::bailOut("Calibration samples removal failed\nExiting\n", 1) unless $remove_ok;
		
		print "Recalculating & setting calibration...\n";
    	$c = new t2tsw::t2tCmd();
    	$c->{cmd} = "START_CALIBRATION";
    	$c->{prm}->{start_calibration}->{load_from_file} = 0;
    	$c->{prm}->{start_calibration}->{cmatrix}->{cols} = 0;
    	$c->{prm}->{start_calibration}->{cmatrix}->{rows} = 0;
    	$c->{prm}->{start_calibration}->{cmatrix}->{vals} = undef;
    	$c->{prm}->{start_calibration}->{fname} = undef;
    	t2tHelpers::cmdEx($c);
    	
    	t2tHelpers::checkStatus('calibrating', 1, 5);
		t2tHelpers::bailOut("Load calibration failed\nExiting\n", 1) unless t2tHelpers::checkStatus('calibstarted', 1, 5);
		t2tHelpers::bailOut("Load calibration failed\nExiting\n", 1) unless t2tHelpers::checkStatus('calibstarted', 0, 5);
    } 
    
    printf "Synchronizing...\n";
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "SYNCHRONISE";
    t2tHelpers::cmdEx($c);
    t2tHelpers::checkStatus('synchronise', 1, 5);
    t2tHelpers::bailOut("Synchronization failed\nExiting\n", 1) unless t2tHelpers::checkStatus('synchronised', 1, 5);
    
    printf "Start auto sync...\n";
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "START_AUTO_SYNC";
    t2tHelpers::cmdEx($c);
    t2tHelpers::bailOut("Start auto sync failed\nExiting\n", 1) unless t2tHelpers::checkStatus('autosynced', 1, 5);
    
    printf "Start tracking...\n";
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "START_TRACKING";
    t2tHelpers::cmdEx($c);
    t2tHelpers::checkStatus('running', 1, 5);
    t2tHelpers::bailOut("Start tracking failed\nExiting\n", 1) unless t2tHelpers::checkStatus('runstarted', 1, 5);
    
    printf "Getting some extended samples...\n";
    for (my $i = 0; $i < 10; $i++) {
    	$c = new t2tsw::t2tCmd();
    	$c->{cmd} = "GET_SAMPLE_EXT";
    	my $smp = t2tHelpers::cmdEx($c)->{sample_ext};
    	printf "> %s\n", $smp->str();
    	Time::HiRes::sleep 0.2;
    }
    
    printf "Getting some timestamps...\n";
    for (my $j = 0; $j < 10; $j++) {
    	$c = new t2tsw::t2tCmd();
    	$c->{cmd} = undef;
    	my $timestamp = t2tHelpers::cmdEx($c)->{timestamp};
    	printf "> %f\n", $timestamp;
    }
    
    printf "Start recording...\n";
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "RECORD";
    t2tHelpers::cmdEx($c);
    Time::HiRes::sleep 2;
	
	print "Sending events...\n";
	for (my $i = 0; $i < 5; $i++) {
    	$c = new t2tsw::t2tCmd();
    	$c->{cmd} = "EVENT";
    	my $name = $c->{prm}->{event}->{name} = "Event $i";
    	my $dur = $c->{prm}->{event}->{duration} = 1;
		$c->{prm}->{event}->{nfields} = 2;
    	my $s = $c->{prm}->{event}->{fields} = t2tsw::new_charpArray(2);
		t2tsw::charpArray_setitem($s, 0, "FIELD 1");
		t2tsw::charpArray_setitem($s, 1, "FIELD 2");
		my $v = $c->{prm}->{event}->{values} = t2tsw::new_doubleArray(2);
		t2tsw::doubleArray_setitem($v, 0, $i*10);
		t2tsw::doubleArray_setitem($v, 1, $i*10+1);
		my @t = Time::HiRes::gettimeofday;
    	my $start = $t[0]+$t[1]/1000000;
		my $evt = t2tHelpers::cmdEx($c);
		t2tsw::delete_doubleArray($v);
		t2tsw::delete_charpArray($s);
		printf "> Name = $name local_start = $start secs real_start = $evt->{start_time} secs duration = $dur secs\n";
        printf "> Sleeping a bit...\n";
    	Time::HiRes::sleep 1.5;
    }
    
    printf "Getting gazes data...\n";
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "GET_GAZES_DATA";
    $c->{prm}->{get_gazes_data}->{from_sample_idx} = 0;
    my $gazes = t2tHelpers::cmdEx($c)->{gazes_data};
    if (defined $gazes) {
    	printf "> Start time: $gazes->{start_time}\n";
    	if (@{$gazes->samples} > 0) {
			for (my $i = 0; $i < @{$gazes->samples}; $i++) {
				printf "> %s\n", $gazes->samples->[$i]->str();
			}
		} else {
			printf "No data\n";
		}
    }
    
    printf "Getting events data...\n";
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "GET_EVENTS_DATA";
    $c->{prm}->{get_events_data}->{from_event_idx} = 0;
    my $ev = t2tHelpers::cmdEx($c)->{events_data};
    if (defined $ev) {
    	printf "> Start time: $ev->{start_time}\n";
    	if (@{$ev->events} > 0) {
			for (my $i = 0; $i < @{$ev->events}; $i++) {
				printf "> %s\n", $ev->events->[$i]->str();
			}
		} else {
			printf "No data\n";
		}
    }
    
    printf "Saving data...\n";
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "SAVE_DATA";
    $c->{prm}->{save_data}->{eye_tracking_fname} = $opt_data_file;   
    $c->{prm}->{save_data}->{events_fname} = $opt_evts_file;
    $c->{prm}->{save_data}->{mode} = "TRUNK";
    t2tHelpers::cmdEx($c);
    
    printf "Getting status and history...\n";
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "GET_STATUS";
    $c->{prm}->{get_status}->{get_history} = 1;
    my $status = t2tHelpers::cmdEx($c);
    printf "> Status: %s\n",  $status->{status_data}->str();
    for (my $i = 0; $i < @{$status->{history_data}->{facts}}; $i++) {
        printf "> %s\n", $status->{history_data}->{facts}->[$i]->str();
    }
    
    printf "Stop recording...\n";
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "STOP_RECORD";
    t2tHelpers::cmdEx($c);
    
    printf "Stop tracking...\n";
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "STOP_TRACKING";
    t2tHelpers::cmdEx($c);
    t2tHelpers::checkStatus('stop', 1, 5);
    t2tHelpers::bailOut("Stop tracking failed\nExiting\n", 0) unless t2tHelpers::checkStatus('running', 0, 5);
    
    printf "Clearing data...\n";
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "CLEAR_DATA";
    $c->{prm}->{clear_data}->{up_sample_idx} = -2;
    $c->{prm}->{clear_data}->{up_event_idx} = -2;
    t2tHelpers::cmdEx($c);
    
    printf "Clearing history...\n";
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "CLEAR_HISTORY";
    t2tHelpers::cmdEx($c);
    
	printf "Stop auto sync...\n";
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "STOP_AUTO_SYNC";
    t2tHelpers::cmdEx($c);
    t2tHelpers::bailOut("Stop auto sync failed\nExiting\n", 1) unless t2tHelpers::checkStatus('autosynced', 0, 5);
    
    $disconnect = 0;
    printf "Disconnecting...\n";
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "DISCONNECT";
    t2tHelpers::cmdEx($c, 0);
    t2tHelpers::checkStatus('disconnect', 1, 5);
	t2tHelpers::bailOut("Disconnection failed\nExiting\n", 0) unless t2tHelpers::checkStatus('connected', 0, 5);
	
	printf "Cleanup...\n";
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "CLEANUP";
    t2tHelpers::cmdEx($c);
    
	printf "End.\n";
	
} or do {
	my $msg = "--------------- UNEXPECTED ERROR ---------------\n$@";
    $msg .= "------------------------------------------------\n";
    $msg .= "The application has been forcely closed!\n";
    t2tHelpers::bailOut($msg, $disconnect);
}

    